From de0df5ad2c59b979db3199018cb00bd6a54df54e Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Tue, 19 Nov 2024 16:44:45 +0100
Subject: [PATCH] usb: dwc2: avoid a race between ERLYSUSP and SUSP interrupt

There are two handlers that may take these interrupts in various order.
This could lead to endless ERLYSUSP interrupt loop, in case SUSP
interrupt is handled first. Normally the ERLYSUSP interrupt should
occur before (e.g. idle state detected after 3ms on the bus), before
the bus suspend state is reached.

When the driver 1st take SUSP interrupt, it will put the controller
into one of the low power modes. For example, when in
DWC2_POWER_DOWN_PARAM_NONE, it can enter clock gating. Then, ERLYSUSP
interrupt handler will fail to clear IRQ flag.

So enforce the order of the handler, when both ERLYSUSP and SUSP are
read: don't clear SUSP flag, simply return. Then the handler in gadget.c
will get called to clear ERLYSUSP. Finally, SUSP interrupt handler will
be called again, so it can be properly handled (go to one of the low
power/suspend modes).

Similarly, some EP interrupts may be pending (not yet handled), when
the bus suspend is seen. This may happen in case the cable gets
unplugged suddenly during enumeration. Give a chance for the gadget
interrupt handler to run, before handling SUSP core interrupt properly.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I7945450784213f1042cb70474bc4ff6a76203d52
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/419639
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/usb/dwc2/core_intr.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/usb/dwc2/core_intr.c b/drivers/usb/dwc2/core_intr.c
index b7577cd5ddae..b9cb876e5952 100644
--- a/drivers/usb/dwc2/core_intr.c
+++ b/drivers/usb/dwc2/core_intr.c
@@ -499,9 +499,25 @@ static void dwc2_handle_disconnect_intr(struct dwc2_hsotg *hsotg)
  */
 static void dwc2_handle_usb_suspend_intr(struct dwc2_hsotg *hsotg)
 {
+	u32 gintsts = dwc2_readl(hsotg, GINTSTS) & dwc2_readl(hsotg, GINTMSK);
 	u32 dsts;
 	int ret;
 
+	if (gintsts & GINTSTS_ERLYSUSP) {
+		dev_dbg(hsotg->dev, "USBSUSP temporarily ignored, raced with ERLYSUSP\n");
+		return;
+	}
+
+	if (gintsts & (GINTSTS_OEPINT | GINTSTS_IEPINT)) {
+		u32 daint = dwc2_readl(hsotg, DAINT);
+		u32 daintmsk = dwc2_readl(hsotg, DAINTMSK);
+
+		if (daint & daintmsk) {
+			dev_dbg(hsotg->dev, "USBSUSP temporarily ignored, before EP IRQ\n");
+			return;
+		}
+	}
+
 	/* Clear interrupt */
 	dwc2_writel(hsotg, GINTSTS_USBSUSP, GINTSTS);
 
-- 
2.50.1

